pkg/server/plugin/noderesolver/aws/iid.go

```diff
@@ -0,0 +1,282 @@
+}
+
+type IIDResolverPlugin struct {
+	mu      sync.RWMutex
+	conf    *aws.Config
+	clients map[string]awsClient
+
+	hooks struct {
+		getenv    func(string) string
+		newClient func(conf *aws.Config) (awsClient, error)
+	}
+}
+
+var _ noderesolver.Plugin = (*IIDResolverPlugin)(nil)
+
+func NewIIDResolverPlugin() *IIDResolverPlugin {
+	p := &IIDResolverPlugin{}
+	p.hooks.getenv = os.Getenv
+	p.hooks.newClient = newAWSClient
+	return p
+}
+
+func (p *IIDResolverPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := new(IIDResolverConfig)
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, iidError.New("unable to decode configuration: %v", err)
+	}
+
+	// Set defaults from the environment
+	if config.AccessKeyID == "" {
+		config.AccessKeyID = p.hooks.getenv("AWS_ACCESS_KEY_ID")
+	}
+	if config.SecretAccessKey == "" {
+		config.SecretAccessKey = p.hooks.getenv("AWS_SECRET_ACCESS_KEY")
+	}
+
+	conf := aws.NewConfig()
+	switch {
+	case config.AccessKeyID != "" && config.SecretAccessKey != "":
+		conf.Credentials = credentials.NewStaticCredentials(config.AccessKeyID, config.SecretAccessKey, "")
+	case config.AccessKeyID != "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing secret access key")
+	case config.AccessKeyID == "" && config.SecretAccessKey != "":
+		return nil, iidError.New("configuration missing access key id")
+	case config.AccessKeyID == "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing both access key id and secret access key")
+	}
+
+	// set the AWS configuration and reset clients
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.conf = conf
+	p.clients = make(map[string]awsClient)
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) Resolve(ctx context.Context, req *noderesolver.ResolveRequest) (*noderesolver.ResolveResponse, error) {
+	resp := &noderesolver.ResolveResponse{
+		Map: make(map[string]*common.Selectors),
+	}
+	for _, spiffeID := range req.BaseSpiffeIdList {
+		selectors, err := p.resolveSpiffeID(ctx, spiffeID)
+		if err != nil {
+			return nil, err
+		}
+		if selectors != nil {
+			resp.Map[spiffeID] = selectors
+		}
+	}
+	return resp, nil
+}
+
+func (p *IIDResolverPlugin) resolveSpiffeID(ctx context.Context, spiffeID string) (*common.Selectors, error) {
+	_, region, instanceID, err := parseAgentID(spiffeID)
+	if err != nil {
+		logrus.Warnf("unrecognized Agent ID: %s: %v", spiffeID, err)
+		return nil, nil
+	}
+
+	client, err := p.getRegionClient(region)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := client.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(instanceID)},
```

Comment:
non-blocking observation: resolver is passed a list of spiffe ids, we could do a single lookup here instead of looping over all of them

Reply:
sure, we could aggregate by region and tackle them at once. i'd rather do that in a follow up PR if that is ok.